Source/Python/callbacks.cpp

```diff
@@ -0,0 +1,52 @@
+/* Copyright 2019-2022 The WarpX Community
+ *
+ * This file is part of WarpX.
+ *
+ * Authors: David Grote, Maxence Thevenet, Weiqun Zhang, Roelof Groenewald, Axel Huebl
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#include "callbacks.H"
+
+#include <cstdlib>
+#include <exception>
+
+
+std::map< std::string, std::function<void()> > warpx_callback_py_map;
+
+void InstallPythonCallback ( std::string name, std::function<void()> callback )
+{
+    warpx_callback_py_map[name] = callback;
+}
+
+bool IsPythonCallbackInstalled ( std::string name )
+{
+    return (warpx_callback_py_map.count(name) == 1u);
+}
+
+// Execute Python callbacks of the type given by the input string
+void ExecutePythonCallback ( std::string name )
+{
+    if ( IsPythonCallbackInstalled(name) ) {
+        WARPX_PROFILE("warpx_py_" + name);
+        try {
+            warpx_callback_py_map[name]();
+        } catch (std::exception &e) {
+            std::cerr << "Python callback '" << name << "' failed!" << std::endl;
+            std::cerr << e.what() << std::endl;
+            std::exit(3);  // note: NOT amrex::Abort(), to avoid hangs with MPI
+
+            // future note:
+            // if we want to rethrow/raise exceptions from Python callbacks through here (C++) and
+            // back the managing Python interpreter, we first need to discard and clear
+            // out the Python error in py::error_already_set. Otherwise, MPI-runs will hang
+            // (and Python will be in continued error state).
+            // https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html#handling-unraisable-exceptions
```

Comment:
I played a bit more with this and the result is not trivial:
Imagine we install into `installbeforestep` and a Python exception is raised. If we rethrow it to Python, we will not finish the step.

The only other potentially useful option I see is to ignore Python errors (w/o rethrow) if we want to. That way, we finish the C++ code that is after the callback.

Reply:
```diff
diff --git a/Examples/Physics_applications/laser_acceleration/PICMI_inputs_3d.py b/Examples/Physics_applications/laser_acceleration/PICMI_inputs_3d.py
index ac3398e43..b044a8801 100755
--- a/Examples/Physics_applications/laser_acceleration/PICMI_inputs_3d.py
+++ b/Examples/Physics_applications/laser_acceleration/PICMI_inputs_3d.py
@@ -1,6 +1,8 @@
 #!/usr/bin/env python3
 
 from pywarpx import picmi
+from pywarpx.callbacks import installbeforestep
+from mpi4py import MPI as mpi
 
 # Physical constants
 c = picmi.constants.c
@@ -159,5 +161,18 @@ sim.write_input_file(file_name = 'inputs_3d_picmi')
 sim.initialize_inputs()
 sim.initialize_warpx()
 
+def mycallback():
+    comm = mpi.COMM_WORLD
+    rank = comm.Get_rank()
+    print(f"I am called ({rank})!")
+    if rank == 0:
+        mpi.oopsi
+
+# callback
+installbeforestep(mycallback)
+
 # Advance simulation until last time step
-sim.step(max_steps)
+try:
+    sim.step(max_steps)
+except RuntimeError:
+    print("ahhh")
diff --git a/Source/Python/callbacks.H b/Source/Python/callbacks.H
index f0c98caf6..fe8c247c5 100644
--- a/Source/Python/callbacks.H
+++ b/Source/Python/callbacks.H
@@ -25,6 +25,7 @@
  * list of currently supported callback names.
 */
 extern WARPX_EXPORT std::map< std::string, std::function<void()> > warpx_callback_py_map;
+extern WARPX_EXPORT std::function<void(std::function<void()>&, std::string const &)> warpx_callback_exec_handler;
 
 /**
  * \brief Function to install the given name and function in warpx_callback_py_map
@@ -39,7 +40,7 @@ bool IsPythonCallbackInstalled ( std::string name );
 /**
  * \brief Function to look for and execute Python callbacks
  */
-void ExecutePythonCallback ( std::string name );
+void ExecutePythonCallback ( std::string const & name );
 
 /**
  * \brief Function to clear the given callback name from warpx_callback_py_map
diff --git a/Source/Python/callbacks.cpp b/Source/Python/callbacks.cpp
index a163f4ac1..f809a7e16 100644
--- a/Source/Python/callbacks.cpp
+++ b/Source/Python/callbacks.cpp
@@ -10,9 +10,31 @@
 
 #include <cstdlib>
 #include <exception>
+#include <iostream>
 
 
 std::map< std::string, std::function<void()> > warpx_callback_py_map;
+std::function<void(std::function<void()>&, std::string const &)> warpx_callback_exec_handler;
+
+namespace {
+    void default_callback_exec_handler (std::function<void()>& callback, std::string const & name)
+    {
+        try {
+            callback();
+        } catch (std::exception &e) {
+            std::cerr << "Python callback '" << name << "' failed!" << std::endl;
+            std::cerr << e.what() << std::endl;
+            std::exit(3);  // note: NOT amrex::Abort(), to avoid hangs with MPI
+
+            // future note:
+            // if we want to rethrow/raise exceptions from Python callbacks through here (C++) and
+            // back the managing Python interpreter, we first need to discard and clear
+            // out the Python error in py::error_already_set. Otherwise, MPI-runs will hang
+            // (and Python will be in continued error state).
+            // https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html#handling-unraisable-exceptions
+        }
+    }
+}
 
 void InstallPythonCallback ( std::string name, std::function<void()> callback )
 {
@@ -25,24 +47,14 @@ bool IsPythonCallbackInstalled ( std::string name )
 }
 
 // Execute Python callbacks of the type given by the input string
-void ExecutePythonCallback ( std::string name )
+void ExecutePythonCallback ( std::string const & name )
 {
     if ( IsPythonCallbackInstalled(name) ) {
         WARPX_PROFILE("warpx_py_" + name);
-        try {
-            warpx_callback_py_map[name]();
-        } catch (std::exception &e) {
-            std::cerr << "Python callback '" << name << "' failed!" << std::endl;
-            std::cerr << e.what() << std::endl;
-            std::exit(3);  // note: NOT amrex::Abort(), to avoid hangs with MPI
-
-            // future note:
-            // if we want to rethrow/raise exceptions from Python callbacks through here (C++) and
-            // back the managing Python interpreter, we first need to discard and clear
-            // out the Python error in py::error_already_set. Otherwise, MPI-runs will hang
-            // (and Python will be in continued error state).
-            // https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html#handling-unraisable-exceptions
-        }
+        if (warpx_callback_exec_handler)
+            warpx_callback_exec_handler(warpx_callback_py_map[name], name);
+        else
+            default_callback_exec_handler(warpx_callback_py_map[name], name);
     }
 }
 
diff --git a/Source/Python/pyWarpX.cpp b/Source/Python/pyWarpX.cpp
index 26f4c7750..405a64dbf 100644
--- a/Source/Python/pyWarpX.cpp
+++ b/Source/Python/pyWarpX.cpp
@@ -10,6 +10,10 @@
 #include <Utils/WarpXUtil.H>  // todo: move to its own Python/Utils.cpp
 #include <Utils/WarpXVersion.H>
 #include <Initialization/WarpXAMReXInit.H>
+#include <Python/callbacks.H>
+
+#include <exception>
+#include <iostream>
 
 #define STRINGIFY(x) #x
 #define MACRO_STRINGIFY(x) STRINGIFY(x)
@@ -28,6 +32,39 @@
 
 //using namespace warpx;
 
+namespace {
+    void pybind_callback_exec_handler (std::function<void()>& callback, std::string const & name)
+    {
+        try {
+            callback();
+        } catch (py::error_already_set &eas) {
+            // if we want to rethrow/raise exceptions from Python callbacks through here (C++) and
+            // back the managing Python interpreter, we first need to discard and clear
+            // out the Python error in py::error_already_set. Otherwise, MPI-runs will hang
+            // (and Python will be in continued error state).
+            // https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html#handling-unraisable-exceptions
+
+            // Python exceptions: print
+            std::cerr << "Python callback '" << name << "' failed!" << std::endl;
+            std::cerr << eas.what() << std::endl;
+
+            // Discard the Python error using Python APIs
+            eas.discard_as_unraisable(__func__);
+
+            // raise as a regular runtime error, which can be caught from Python
+            //   downside: C++ code that is written after the callback hook will not be executed
+            // throw std::runtime_error("Python callback '" + name + "' failed!");
+
+            // do NOT abort but return to Python
+        } catch (std::exception &e) {
+            // C++ and all other exceptions: print and discard
+            std::cerr << "Python callback '" << name << "' failed!" << std::endl;
+            std::cerr << e.what() << std::endl;
+
+            // do NOT abort but return to Python
+        }
+    }
+}
 
 // forward declarations of exposed classes
 void init_BoundaryBufferParIter (py::module&);
@@ -58,6 +95,8 @@ PYBIND11_MODULE(PYWARPX_MODULE_NAME, m) {
                WarpX
     )pbdoc";
 
+    warpx_callback_exec_handler = pybind_callback_exec_handler;
+
     // note: order from parent to child classes
     init_PinnedMemoryParticleContainer(m);
     init_WarpXParticleContainer(m);
```